crates/kubelet/src/state.rs

```diff
@@ -1,25 +1,221 @@
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// Edges must be defined, even for self-transition, with edge removed, compilation fails:
+    /// ```compile_fail
+    /// use kubelet::state::{Transition, State};
+    /// use kubelet::pod::Pod;
+    ///
+    /// #[derive(Debug)]
+    /// struct TestState;
+    ///
+    /// // impl EdgeTo<TestState> for TestState {}
+    ///
+    /// struct PodState;
+    ///
+    /// #[async_trait::async_trait]
+    /// impl State<PodState> for TestState {
+    ///     async fn next(
+    ///         &self,
+    ///         _pod_state: &mut PodState,
+    ///         _pod: &Pod,
+    ///     ) -> anyhow::Result<Transition<PodState>> {
+    ///         // This fails because TestState is not EdgeTo<TestState>
+    ///         Ok(Transition::next(self, TestState))
+    ///     }
+    ///
+    ///     async fn json_status(
+    ///         &self,
+    ///         _pod_state: &mut PodState,
+    ///         _pod: &Pod,
+    ///     ) -> anyhow::Result<serde_json::Value> {
+    ///         Ok(serde_json::json!(null))
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// The next state must have the same PodState type, otherwise compilation will fail:
+    /// ```compile_fail
+    /// use kubelet::state::{Transition, State, EdgeTo};
+    /// use kubelet::pod::Pod;
+    ///
+    /// #[derive(Debug)]
+    /// struct TestState;
+    ///
+    /// struct PodState;
+    ///
+    /// #[derive(Debug)]
+    /// struct OtherState;
+    ///
+    /// struct OtherPodState;
+    ///
+    /// impl EdgeTo<OtherState> for TestState {}
+    ///
+    /// #[async_trait::async_trait]
+    /// impl State<PodState> for TestState {
+    ///     async fn next(
+    ///         &self,
+    ///         _pod_state: &mut PodState,
+    ///         _pod: &Pod,
+    ///     ) -> anyhow::Result<Transition<PodState>> {
+    ///         // This fails because OtherState is State<OtherPodState>
+    ///         Ok(Transition::next(self, OtherState))
+    ///     }
+    ///
+    ///     async fn json_status(
+    ///         &self,
+    ///         _pod_state: &mut PodState,
+    ///         _pod: &Pod,
+    ///     ) -> anyhow::Result<serde_json::Value> {
+    ///         Ok(serde_json::json!(null))
+    ///     }
+    /// }
+    ///
+    /// #[async_trait::async_trait]
+    /// impl State<OtherPodState> for OtherState {
+    ///     async fn next(
+    ///         &self,
+    ///         _pod_state: &mut OtherPodState,
+    ///         _pod: &Pod,
+    ///     ) -> anyhow::Result<Transition<OtherPodState>> {
+    ///         Ok(Transition::Complete(Ok(())))
+    ///     }
+    ///
+    ///     async fn json_status(
+    ///         &self,
+    ///         _pod_state: &mut OtherPodState,
+    ///         _pod: &Pod,
+    ///     ) -> anyhow::Result<serde_json::Value> {
+    ///         Ok(serde_json::json!(null))
+    ///     }
+    /// }
+    /// ```
+    pub fn next<I: State<PodState>, S: State<PodState>>(_i: &I, s: S) -> Transition<PodState>
```

Comment:
This feels a little weird to make the user pass the old state in order to enforce the edge constraint. Is there any other way around this? I can't think of any and I don't know what you already tried

Reply:
I spent quite a while to try to avoid this but couldn’t come up with anything. Happy to remove it if we can come up with another way, but I don’t think it’s possible since this is a class method and has no context.